osf/models/collection_submission.py

```diff
@@ -0,0 +1,255 @@
+            raise PermissionsError(f'{user} must have admin permissions.')
+
+        is_admin = self.guid.referent.has_permission(user, ADMIN)
+        if not is_admin:
+            raise PermissionsError(f'{user} must have admin permissions.')
+
+    def _save_transition(self, event_data):
+        '''Save changes here and write the action.'''
+        self.save()
+        from_state = CollectionSubmissionStates[event_data.transition.source]
+        to_state = self.state
+
+        trigger = CollectionSubmissionsTriggers.from_db_name(event_data.event.name)
+        if trigger is None:
+            return
+
+        self.actions.create(
+            from_state=from_state,
+            to_state=to_state,
+            trigger=trigger,
+            creator=event_data.kwargs.get('user', self.creator),
+            comment=event_data.kwargs.get('comment', '')
+        )
+
+    @cached_property
+    def _id(self):
+        return '{}-{}'.format(self.guid._id, self.collection._id)
+
+    @classmethod
+    def load(cls, data, select_for_update=False):
+        try:
+            collection_submission_id, collection_id = data.split('-')
+        except ValueError:
+            raise ValueError('Invalid CollectionSubmission object <_id {}>'.format(data))
+        else:
+            if collection_submission_id and collection_id:
+                try:
+                    if isinstance(data, basestring):
+                        return (cls.objects.get(guid___id=collection_submission_id, collection__guids___id=collection_id) if not select_for_update
+                                else cls.objects.filter(guid___id=collection_submission_id, collection__guids___id=collection_id).select_for_update().get())
+                except cls.DoesNotExist:
+                    return None
+            return None
+
+    @property
+    def absolute_api_v2_url(self):
+        path = '/collections/{}/collection_submissions/{}/'.format(self.collection._id, self.guid._id)
+        return api_v2_url(path)
+
+    def update_index(self):
+        if self.collection.is_public:
+            from website.search.search import update_collected_metadata
+            try:
+                update_collected_metadata(self.guid._id, collection_id=self.collection.id)
+            except SearchUnavailableError as e:
+                logger.exception(e)
+
+    def remove_from_index(self):
+        from website.search.search import update_collected_metadata
+        try:
+            update_collected_metadata(self.guid._id, collection_id=self.collection.id, op='delete')
+        except SearchUnavailableError as e:
+            logger.exception(e)
+
+    def save(self, *args, **kwargs):
+        kwargs.pop('old_subjects', None)  # Not indexing this, trash it
+        ret = super(CollectionSubmission, self).save(*args, **kwargs)
+        self.update_index()
+        return ret
+
+    def delete(self, *args, **kwargs):
+        self.remove_from_index()
+        super().delete()
```

Comment:
still need to know if we want a `deleted` field on CollectionSubmissions or if we just want to treat `REMOVED` state the same way